dask_searchcv/tests/test_adaptive.py

```diff
@@ -0,0 +1,181 @@
+from dask_searchcv.adaptive import _top_k, Hyperband
+from dask_ml.linear_model import PartialSGDClassifier
+from dask_ml.datasets import make_classification
+import numpy as np
+import dask.array as da
+from dask.distributed import Client
+import pandas as pd
+import pytest
+import distributed
+from pprint import pprint
+import scipy.stats as stats
+import random
+from sklearn.linear_model import Lasso
+from distributed.utils_test import cluster, loop, gen_cluster
+import time
+from dask_ml.wrappers import Incremental
+
+
+class ConstantFunction:
+    def _fn(self):
+        return self.value
+
+    def get_params(self, deep=None, **kwargs):
+        return {k: getattr(self, k) for k, v in kwargs.items()}
+
+    def set_params(self, **kwargs):
+        for k, v in kwargs.items():
+            setattr(self, k, v)
+        return self
+
+    def partial_fit(self, *args, **kwargs):
+        pass
+
+    def score(self, *args, **kwargs):
+        return self._fn()
+
+
+def _with_client(fn):
+    def fn_with_client(*args, **kwargs):
+        with cluster() as (s, [a, b]):
+            with Client(s['address'], loop=loop) as client:
+                y = fn(*args, **kwargs)
+        return y
+    return fn_with_client
+
+
+#  @_with_client
+@gen_cluster(client=True)
+def test_hyperband_sklearn(*args, **kwargs):
```

Comment:
Doepending on the options to `gen_cluster`, the test is called with the following

- client
- scheduler
- list of workers,

so the typical function signature would be 

```

@gen_cluster(client=True)
def test_hyperband_sklearn(c, s, a, b): 
    ...
```

Reply:
That said, (I think) `gen_cluster` tests require essentially everything to be async. After import `sklearn.linear_model.SGDClassifier`, this fails with

```
../distributed/distributed/utils_test.py:808:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../Envs/dask-dev/lib/python3.6/site-packages/tornado/ioloop.py:582: in run_sync
    return future_cell[0].result()
../../Envs/dask-dev/lib/python3.6/site-packages/tornado/gen.py:1113: in run
    yielded = self.gen.send(value)
../distributed/distributed/utils_test.py:789: in coro
    result = yield future
../../Envs/dask-dev/lib/python3.6/site-packages/tornado/gen.py:1099: in run
    value = future.result()
../../Envs/dask-dev/lib/python3.6/site-packages/tornado/gen.py:296: in wrapper
    result = func(*args, **kwargs)
/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/types.py:248: in wrapped
    coro = func(*args, **kwargs)
dask_searchcv/tests/test_adaptive.py:50: in test_hyperband_sklearn
    X, y = make_classification(n_samples=1000, chunks=500)
../dask-ml/dask_ml/datasets.py:332: in make_classification
    informative_idx, beta = dask.compute(informative_idx, beta)
../dask/dask/base.py:401: in compute
    return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

.0 = <zip object at 0x116259a08>

>   return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])
E   AssertionError: yield from wasn't used with future
```

The problem is the line `informative_idx, beta = dask.compute(informative_idx, beta)` coming from `dask_ml.datasets`.

So I'd recommend trying a `with cluster` style test.

Reply:
So for this diff

```diff
diff --git a/dask_searchcv/tests/test_adaptive.py b/dask_searchcv/tests/test_adaptive.py
index 199c5a1..2fbe5e8 100644
--- a/dask_searchcv/tests/test_adaptive.py
+++ b/dask_searchcv/tests/test_adaptive.py
@@ -1,18 +1,15 @@
 from dask_searchcv.adaptive import _top_k, Hyperband
-from dask_ml.linear_model import PartialSGDClassifier
+from sklearn.linear_model import SGDClassifier
 from dask_ml.datasets import make_classification
 import numpy as np
 import dask.array as da
 from dask.distributed import Client
 import pandas as pd
 import pytest
-import distributed
 from pprint import pprint
 import scipy.stats as stats
-import random
 from sklearn.linear_model import Lasso
 from distributed.utils_test import cluster, loop, gen_cluster
-import time
 from dask_ml.wrappers import Incremental
 
 
@@ -45,23 +42,24 @@ def _with_client(fn):
 
 
 #  @_with_client
-@gen_cluster(client=True)
-def test_hyperband_sklearn(*args, **kwargs):
-    X, y = make_classification(n_samples=1000, chunks=500)
-    classes = da.unique(y)
-    model = Incremental(SGDClassifier(warm_start=True, classes=classes,
-                                      loss='hinge', penalty='elasticnet'))
-
-    params = {'alpha': np.logspace(-3, 0, num=int(10e3)),
-              'l1_ratio': np.linspace(0, 1, num=int(10e3))}
-    alg = Hyperband(model, params, max_iter=9, n_jobs=0)
-
-    alg.fit(X, y, dry_run=True)
-    assert len(alg.history) == 20
-    alg.fit(X, y, dry_run=True)
-    assert len(alg.history) == 40
-
-    alg.fit(X, y, classes=classes)
+def test_hyperband_sklearn():
+    with cluster() as (s, (a, b)):
+        client = Client(s['address'])  # noqa
+        X, y = make_classification(n_samples=1000, chunks=500)
+        classes = da.unique(y)
+        model = Incremental(SGDClassifier(warm_start=True,
+                                          loss='hinge', penalty='elasticnet'))
+
+        params = {'alpha': np.logspace(-3, 0, num=int(10e3)),
+                  'l1_ratio': np.linspace(0, 1, num=int(10e3))}
+        alg = Hyperband(model, params, max_iter=9, n_jobs=0)
+
+        alg.fit(X, y, dry_run=True, classes=[0, 1])
+        assert len(alg.history) == 20
+        alg.fit(X, y, dry_run=True, classes=[0, 1])
+        assert len(alg.history) == 40
+
+        alg.fit(X, y, classes=classes)
 
 
 #  @_with_client
```

That's running, but taking a while. Not sure if it's a long test, or if it's deadlocked somewhere

Reply:
Yeah, I'm resolving some of these issues now.